datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java

```diff
@@ -141,6 +142,126 @@ public StickyPartitionAssignmentStrategy(Optional<Integer> maxTasks, Optional<In
+        targetAssignment, partitionsMetadata.getPartitions());
+
+    DatastreamGroup dg = partitionsMetadata.getDatastreamGroup();
+
+    Set<String> allToReassignPartitions = new HashSet<>();
+    targetAssignment.values().forEach(allToReassignPartitions::addAll);
+    allToReassignPartitions.retainAll(partitionsMetadata.getPartitions());
+
+    // construct a map to store the tasks and if it contain the partitions that can be released
+    // map: <source taskName, partitions that need to be released>
+    Map<String, Set<String>> confirmedPartitionsTaskMap = new HashMap<>();
+
+    // construct a map to store the partition and its source task
+    // map: <partitions that need to be released, source taskName>
+    Map<String, String> partitionToSourceTaskMap = new HashMap<>();
+
+    // We first confirm that the partitions in the target assignment which can be removed, and we find out its source task
+    // If the partitions cannot be found from any task, we ignore these partitions
+    currentAssignment.keySet().forEach(instance -> {
+      Set<DatastreamTask> tasks = currentAssignment.get(instance);
+      tasks.forEach(task -> {
+        if (dg.belongsTo(task)) {
+          Set<String> toMovePartitions = new HashSet<>(task.getPartitionsV2());
+          toMovePartitions.retainAll(allToReassignPartitions);
+          confirmedPartitionsTaskMap.put(task.getDatastreamTaskName(), toMovePartitions);
+          toMovePartitions.forEach(p -> partitionToSourceTaskMap.put(p, task.getDatastreamTaskName()));
+        }
+      });
+    });
+
+    Set<String> tasksToMutate = confirmedPartitionsTaskMap.keySet();
+    Set<String> toReleasePartitions = new HashSet<>();
+    confirmedPartitionsTaskMap.values().forEach(v -> toReleasePartitions.addAll(v));
+
+    // Compute new assignment from the current assignment
+    Map<String, Set<DatastreamTask>> newAssignment = new HashMap<>();
+
+    currentAssignment.keySet().forEach(instance -> {
+      Set<DatastreamTask> tasks = currentAssignment.get(instance);
+
+      // check if this instance has any partition to be added
+      final Set<String> toAddedPartitions = new HashSet<>();
+      if (targetAssignment.containsKey(instance)) {
+        // filter the target assignment by the partitions which have a confirmed source
+        Set<String> p = targetAssignment.get(instance).stream()
+            .filter(toReleasePartitions::contains).collect(Collectors.toSet());
+        toAddedPartitions.addAll(p);
+      }
+
+      Set<DatastreamTask> dgTasks = tasks.stream().filter(dg::belongsTo)
+          .collect(Collectors.toSet());
+      if (toAddedPartitions.size() > 0 && dgTasks.isEmpty()) {
+        String errorMsg = String.format("No task is available in the target instance %s", instance);
+        LOG.error(errorMsg);
+        throw new DatastreamRuntimeException(errorMsg);
+      }
+
+      // find the task with minimum number of partitions on that instance to store the moved partitions
```

Comment:
Instead of assigning all the partitions to just one task, is it possible to distribute it amongst all the tasks on that host, with tasks with less# partitions getting more of new partitions? I am thinking of a corner case where a lot of partitions are moved to one host, and they all get assigned to just one task.

Reply:
Yes, this is possible. However, I'm inclined to keep the current algorithm primary due to 1) less mutation tasks which may also affect performance 2) simplicity in the algorithm 3) easy to control the assignment, for example we can decide to put into one task by a single call, or put into possible two tasks by splitting into more calls.